community/security/src/main/java/org/neo4j/server/security/auth/RateLimitedAuthenticationStrategy.java

```diff
@@ -87,14 +87,27 @@ public AuthenticationResult authenticate( User user, String password)
         }
     }
 
-    private AuthenticationMetadata authMetadataFor( String username )
+    private AuthenticationMetadata authMetadataFor( String username, String source )
     {
-        AuthenticationMetadata authMeta = authenticationData.get( username );
+        ConcurrentMap<String, AuthenticationMetadata> authMetaMap = authenticationData.get( username );
+
+        if ( authMetaMap == null )
+        {
+            authMetaMap = new ConcurrentHashMap<>();
+            ConcurrentMap<String,AuthenticationMetadata> preExisting = authenticationData.putIfAbsent( username, authMetaMap );
+
+            if ( preExisting != null )
+            {
+                authMetaMap = preExisting;
+            }
+        }
```

Comment:
Instead of this dance, how about `ConcurrentMap.computeIfAbsent`? That would replace both `get` and `putIfAbsent` and guarantees concurrently nice behaviour.